### ISSUE TITLE
Age estimation for faces

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#377]
After testing, improving and merging #361, we can store faces bounding boxes and run some age estimation algorithm like https://www.pyimagesearch.com/2020/04/13/opencv-age-detection-with-deep-learning/

The ImageViewer could also be extended to draw faces bounding boxes stored in items and display ages.

This will be very useful for CSAM cases.

**Comment by hauck-jvsh** [comment]
This will be great.

**Comment by hauck-jvsh** [comment]
Other reference: https://www.pyimagesearch.com/2020/04/13/opencv-age-detection-with-deep-learning/

**Comment by lfcnassif** [comment]
Dataset for testing: https://data.vision.ee.ethz.ch/cvl/rrothe/imdb-wiki/

**Comment by lfcnassif** [comment]
Well, that is worse than I expected. Could you measure the results using precision x recall? I think we should target high recall values, considering CSAM cases.

**Comment by lfcnassif** [comment]
Maybe @joaomacedo could give some advice here.

**Comment by lfcnassif** [comment]
Removing the assignee as this is frozen.

**Comment by lfcnassif** [comment]
Just found this popular model using a Vision Transformer for age prediction:
https://huggingface.co/nateraw/vit-age-classifier

**Comment by lfcnassif** [comment]
Another one, more popular, reporting up to 79% accuracy for <= 9 years old on the tested data set, with 65M downloads/month:
https://huggingface.co/dima806/fairface_age_image_detection

**Comment by lfcnassif** [comment]
Evaluation from NIST:
https://pages.nist.gov/frvt/html/frvt_age_estimation.html

**Comment by lfcnassif** [comment]
This one reports 96% precision and 96% recall for 0-10 years old on the used data set, the higher across all ranges, with more samples for that range, seems to have focused on children:
https://huggingface.co/prithivMLmods/facial-age-detection

**Comment by lfcnassif** [comment]
Another model from the same author above, with better numbers (but less classes) trained on 20K images, the previous one used 10K:
https://huggingface.co/prithivMLmods/open-age-detection

**Comment by lfcnassif** [comment]
Just found this paper:
https://www.sciencedirect.com/science/article/pii/S2666281721000160

I met Dixon at Interpol, he knows a lot about deeplearning and works hard to protect children. They argue they achieved MAE of just 2.36 years in paper above. Code is at:
https://github.com/4ND4/Vec2UAge

**Comment by lfcnassif** [comment]
In a test here, seems this new feature entered in a deadlock or starvation:
<img width="1041" height="991" alt="Image" src="https://github.com/user-attachments/assets/14de05ab-fe85-47d4-91ef-849e0ac8b839" />

Processing is running for dozens of hours in a case that usually finishes in 3h. Processing above is resuming (--continue) a processing that aborted due to disk full and it remained paused for several hours because I needed to use the machine for another heavy task.

**Comment by wladimirleite** [comment]
That happens a couple of times in the last days here, processing real cases.
I didn't have time to analyze it, but taking a quick look at the code, it seems that if "process" and "sendToNext" are not called sequentially, as it may happen for carved items, the batched items control may fail. As it is a Python task, not sure if there some kind of synchronization outside the task code itself...

**Comment by lfcnassif** [comment]
Thanks for reporting @wladimirleite, so I'm reopening this until it is reproduced and fixed. Could you take a look on this @marcosammoura?

I wonder if we could disable batch processing by default, since it usually doesn't help processing on CPU, which is the default.

**Comment by marcosammoura** [comment]
@lfcnassif, what do you think about changing `sendToNextTask` in a similar way you've changed this function on `NSFWNudityDetectTask` via commit [10dce0c](https://github.com/sepinf-inc/IPED/commit/10dce0c4373a41d29d102722cfc492deadbdd638)? I've created commit [0bf4121](https://github.com/sepinf-inc/IPED/commit/0bf412185cf2be0ec6ac4d1a1dbc675e0f7df4de) with the proposed change.

PS: I was not able to reproduce the problem described, so I was not able to test the proposed approach.

**Comment by hauck-jvsh** [comment]
> Thanks for reporting [@wladimirleite](https://github.com/wladimirleite), so I'm reopening this until it is reproduced and fixed. Could you take a look on this [@marcosammoura](https://github.com/marcosammoura)?
> 
> I wonder if we could disable batch processing by default, since it usually doesn't help processing on CPU, which is the default.

This also happens to me, It stays in that loop for all the weekend, and was also after a continue.

**Comment by hauck-jvsh** [comment]
Just to let you all known I was processing this without carving, so it can happen even without it.

**Comment by marcosammoura** [comment]
Could someone rerun a failed case with the change I've suggested above?

**Comment by lfcnassif** [comment]
I'm processing the case again here, but ideally we should be able to reproduce the issue consistently to verify the fix, it happened to me just once.

**Comment by lfcnassif** [comment]
Commit 0bf412185cf2be0ec6ac4d1a1dbc675e0f7df4de didn't fix it:
<img width="1101" height="952" alt="Image" src="https://github.com/user-attachments/assets/90a24719-89de-4602-aaf1-f57b5c93d5b5" />

I've aborted and resumed the processing to reproduce the same conditions, but I don't know if it is related. Did you resumed your processings @wladimirleite?

**Comment by wladimirleite** [comment]
> I've aborted and resumed the processing to reproduce the same conditions, but I don't know if it is related. Did you resumed your processings [@wladimirleite](https://github.com/wladimirleite)?

No, I didn't use "--continue" here. It was the "initial" processing.

**Comment by marcosammoura** [comment]
> Commit [0bf4121](https://github.com/sepinf-inc/IPED/commit/0bf412185cf2be0ec6ac4d1a1dbc675e0f7df4de) didn't fix it: 
> 
> I've aborted and resumed the processing to reproduce the same conditions, but I don't know if it is related. Did you resumed your processings [@wladimirleite](https://github.com/wladimirleite)?

Ok. Trying to figure out the bug yet.

**Comment by lfcnassif** [comment]
@marcosammoura, I just pushed a possible fix. I'm testing it here, but I'm not so confident it will fix the issue.

Based on the screenshots I shared, seems some queued items are being lost (or sent multiple times). When workers detect some item is still under processing (sent to processing pipeline, but not finished yet, maybe queued), they send the queue-end special item to signal all tasks any queued item should be finished.

**Comment by lfcnassif** [comment]
Just added some logging locally to confirm the hypothesis above.

**Comment by marcosammoura** [comment]
> [@marcosammoura](https://github.com/marcosammoura), I just pushed a possible fix. I'm testing it here, but I'm not so confident it will fix the issue.
> 
> Based on the screenshots I shared, seems some queued items are being lost (or sent multiple times). When workers detect some item is still under processing (sent to processing pipeline, but not finished yet, maybe queued), they send the queue-end special item to signal all tasks any queued item should be finished.

> Just added some logging locally to confirm the hypothesis above.

Thanks for this new approach to try to solve the issue. Waiting for the results.

**Comment by lfcnassif** [comment]
Possibly commit [4d27748](https://github.com/sepinf-inc/IPED/commit/4d27748a7851f1a91ce3329b552bd50823e79c40) fixed the issue, in the last 2 runs, processing finished fine. Before the change, I reproduced the issue 2x sequentially. I'm running a third test before opening a PR.

But if @wladimirleite or @hauck-jvsh could test the fix on at least one case that triggered the issue, it would be very welcomed.

**Comment by lfcnassif** [comment]
@wladimirleite and @hauck-jvsh, if you test master on some case that triggered the batched items issue shown in screenshots above, please let us know the results, thanks.

**Comment by wladimirleite** [comment]
I am processing the case that I had issues before (although they didn't happen every time). I will update here if it happens with the modified version.
